File: backend/pkg/apiclient/meal_plan_grocery_list_items.go
package apiclient

import (
	"context"

	"github.com/dinnerdonebetter/backend/internal/observability"
	"github.com/dinnerdonebetter/backend/internal/observability/keys"
	"github.com/dinnerdonebetter/backend/internal/observability/tracing"
	"github.com/dinnerdonebetter/backend/pkg/types"
)

// GetMealPlanGroceryListItem gets a meal plan grocery list item.
func (c *Client) GetMealPlanGroceryListItem(ctx context.Context, mealPlanID, mealPlanGroceryListItemID string) (*types.MealPlanGroceryListItem, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealPlanID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	if mealPlanGroceryListItemID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanGroceryListItemIDKey, mealPlanGroceryListItemID)
	tracing.AttachToSpan(span, keys.MealPlanGroceryListItemIDKey, mealPlanGroceryListItemID)

	req, err := c.requestBuilder.BuildGetMealPlanGroceryListItemRequest(ctx, mealPlanID, mealPlanGroceryListItemID)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "building get meal plan grocery list item request")
	}

	var apiResponse *types.APIResponse[*types.MealPlanGroceryListItem]
	if err = c.fetchAndUnmarshal(ctx, req, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meal plan grocery list item")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// GetMealPlanGroceryListItemsForMealPlan gets a meal plan grocery list item.
func (c *Client) GetMealPlanGroceryListItemsForMealPlan(ctx context.Context, mealPlanID string) ([]*types.MealPlanGroceryListItem, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealPlanID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	req, err := c.requestBuilder.BuildGetMealPlanGroceryListItemsForMealPlanRequest(ctx, mealPlanID)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "building get meal plan grocery list item request")
	}

	var apiResponse *types.APIResponse[[]*types.MealPlanGroceryListItem]
	if err = c.fetchAndUnmarshal(ctx, req, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meal plan grocery list item")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// CreateMealPlanGroceryListItem creates a meal plan grocery list item.
func (c *Client) CreateMealPlanGroceryListItem(ctx context.Context, mealPlanID string, input *types.MealPlanGroceryListItemCreationRequestInput) (*types.MealPlanGroceryListItem, error) {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealPlanID == "" {
		return nil, ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	if input == nil {
		return nil, ErrInvalidIDProvided
	}

	req, err := c.requestBuilder.BuildCreateMealPlanGroceryListItemRequest(ctx, mealPlanID, input)
	if err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "building get meal plan grocery list item request")
	}

	var apiResponse *types.APIResponse[*types.MealPlanGroceryListItem]
	if err = c.fetchAndUnmarshal(ctx, req, &apiResponse); err != nil {
		return nil, observability.PrepareAndLogError(err, logger, span, "retrieving meal plan grocery list item")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return nil, err
	}

	return apiResponse.Data, nil
}

// UpdateMealPlanGroceryListItem updates a meal plan grocery list item.
func (c *Client) UpdateMealPlanGroceryListItem(ctx context.Context, mealPlanID, mealPlanGroceryListItemID string, input *types.MealPlanGroceryListItemUpdateRequestInput) error {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if input == nil {
		return ErrNilInputProvided
	}

	if mealPlanID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	if err := input.ValidateWithContext(ctx); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "validating input")
	}

	req, err := c.requestBuilder.BuildUpdateMealPlanGroceryListItemRequest(ctx, mealPlanID, mealPlanGroceryListItemID, input)
	if err != nil {
		return observability.PrepareAndLogError(err, logger, span, "building create meal plan grocery list item request")
	}

	var apiResponse *types.APIResponse[*types.MealPlanGroceryListItem]
	if err = c.fetchAndUnmarshal(ctx, req, &apiResponse); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "creating meal plan grocery list item")
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return err
	}

	return nil
}

// ArchiveMealPlanGroceryListItem updates a meal plan grocery list item.
func (c *Client) ArchiveMealPlanGroceryListItem(ctx context.Context, mealPlanID, mealPlanGroceryListItemID string) error {
	ctx, span := c.tracer.StartSpan(ctx)
	defer span.End()

	logger := c.logger.Clone()

	if mealPlanID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanIDKey, mealPlanID)
	tracing.AttachToSpan(span, keys.MealPlanIDKey, mealPlanID)

	if mealPlanGroceryListItemID == "" {
		return ErrInvalidIDProvided
	}
	logger = logger.WithValue(keys.MealPlanGroceryListItemIDKey, mealPlanGroceryListItemID)
	tracing.AttachToSpan(span, keys.MealPlanGroceryListItemIDKey, mealPlanGroceryListItemID)

	req, err := c.requestBuilder.BuildArchiveMealPlanGroceryListItemRequest(ctx, mealPlanID, mealPlanGroceryListItemID)
	if err != nil {
		return observability.PrepareAndLogError(err, logger, span, "building create meal plan grocery list item request")
	}

	var apiResponse *types.APIResponse[*types.MealPlanGroceryListItem]
	if err = c.fetchAndUnmarshal(ctx, req, &apiResponse); err != nil {
		return observability.PrepareAndLogError(err, logger, span, "archiving meal plan %s", mealPlanID)
	}

	if err = apiResponse.Error.AsError(); err != nil {
		return err
	}

	return nil
}

File: frontend/packages/api-client/meal_plan_grocery_list_items.ts
import { Axios } from 'axios';
import format from 'string-format';

import {
  MealPlanGroceryListItemCreationRequestInput,
  MealPlanGroceryListItem,
  MealPlanGroceryListItemUpdateRequestInput,
  APIResponse,
} from '@dinnerdonebetter/models';

import { backendRoutes } from './routes';

export async function createMealPlanGroceryListItem(
  client: Axios,
  mealPlanID: string,
  input: MealPlanGroceryListItemCreationRequestInput,
): Promise<MealPlanGroceryListItem> {
  return new Promise(async function (resolve, reject) {
    const response = await client.post<APIResponse<MealPlanGroceryListItem>>(
      format(backendRoutes.MEAL_PLAN_GROCERY_LIST_ITEMS, mealPlanID),
      input,
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getMealPlanGroceryListItem(client: Axios, mealPlanID: string): Promise<MealPlanGroceryListItem> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<MealPlanGroceryListItem>>(
      format(backendRoutes.MEAL_PLAN_GROCERY_LIST_ITEM, mealPlanID),
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function getMealPlanGroceryListItems(
  client: Axios,
  mealPlanID: string,
): Promise<MealPlanGroceryListItem[]> {
  return new Promise(async function (resolve, reject) {
    const response = await client.get<APIResponse<MealPlanGroceryListItem[]>>(
      format(backendRoutes.MEAL_PLAN_GROCERY_LIST_ITEMS, mealPlanID),
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function updateMealPlanGroceryListItem(
  client: Axios,
  mealPlanID: string,
  mealPlanGroceryListItemID: string,
  input: MealPlanGroceryListItemUpdateRequestInput,
): Promise<MealPlanGroceryListItem> {
  return new Promise(async function (resolve, reject) {
    const response = await client.put<APIResponse<MealPlanGroceryListItem>>(
      format(backendRoutes.MEAL_PLAN_GROCERY_LIST_ITEM, mealPlanID, mealPlanGroceryListItemID),
      input,
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}

export async function deleteMealPlanGroceryListItem(
  client: Axios,
  mealPlanID: string,
  mealPlanGroceryListItemID: string,
): Promise<MealPlanGroceryListItem> {
  return new Promise(async function (resolve, reject) {
    const response = await client.delete<APIResponse<MealPlanGroceryListItem>>(
      format(backendRoutes.MEAL_PLAN_GROCERY_LIST_ITEM, mealPlanID, mealPlanGroceryListItemID),
    );

    if (response.data.error) {
      reject(new Error(response.data.error.message));
    }

    resolve(response.data.data);
  });
}
